New: no-restricted-globals rule implementation (fixes #3966)

BenoitZugmeyer opened the pull request:
Implementation of the new rule no-restricted-globals to disallow specific global variables usage. As explained in #3966, some globals defined by environments may be unwanted, so here is a rule to make sure they aren't used.

I have many usages in mind:
* prevent the usage of the deprecated `event` variable defined by the `"browser"` environment
* force the usage of a non native `Promise` implementation even if it is defined in the environment
* prevent the inadvertent usage of `fdescribe` or `fit` in a `"jasmine"` environment

The implementation is pretty straightforward. It is done in two steps: first, check variables defined outside of the current global scope (ex: made global via the eslint configuration), then check variables not defined at all.

platinumazure commented on tests/lib/rules/no-restricted-globals.js:
```diff
@@ -0,0 +1,64 @@
+        { code: "event", env: { browser: true }, options: ["bar"] },
+        { code: "import foo from 'bar';", options: ["foo"], parserOptions: { ecmaVersion: 6, sourceType: "module" } },
+        { code: "function foo() {}", options: ["foo"] },
+        { code: "function fn() { var foo; }", options: ["foo"] }
```
I admit this is a pretty obvious case and not directly related to the rule, but maybe we could use a negative valid test on restricted variables being properties? (E.g., `foo.bar` should be valid when "bar" is specified.)

nzakas commented on docs/rules/no-restricted-globals.md:
```diff
@@ -0,0 +1,50 @@
+Disallowing usage of specific global variables can be useful if you want to allow a set of global
+variables by enabling an environment, but still want to disallow some of those.
+
+For instance, the `"browser"` environment defines `event` as a global variable, but using this
```
Can you make this more generic? The browser environment may change, and this would then be out of date. Maybe talk about `event` without mentioning the environment (say early IE had this).

nzakas commented on lib/rules/no-restricted-globals.js:
```diff
@@ -0,0 +1,70 @@
+/**
+ * @fileoverview Restrict usage of specified globals.
+ * @author Benoît Zugmeyer
```
Add copyright and license info http://eslint.org/docs/developer-guide/code-conventions.html#file-format

nzakas commented on lib/rules/no-restricted-globals.js:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    return {
+        "Program:exit": function() {
```
I don't think ":exit" is necessary here.

nzakas commented on tests/lib/rules/no-restricted-globals.js:
```diff
@@ -0,0 +1,64 @@
+/**
+ * @fileoverview Tests for no-restricted-globals.
+ * @author Benoît Zugmeyer
+ */
```
Please add copyright and license info

nzakas commented on lib/rules/no-restricted-globals.js:
```diff
@@ -0,0 +1,70 @@
+     * @returns {void}
+     * @private
+     */
+    function reportReference(reference) {
```
Can we simplify this message to "Unexpected use of '{{name}}'"?

BenoitZugmeyer commented:
Thanks for the review. I updated the PR to address all points:
* Fix copyrights in .js headers
* Add a test to make sure properties with the same name as a restricted global are still valid
* Rephrased the rule error message
* Rephrased the usage example in documentation
* Changed the `Program:exit` callback to `Program`

nzakas commented:
Lgtm. Thanks!
